ract/cubes/_gpio_impl.py

```diff
@@ -13,25 +13,53 @@
 
 
 class GpioCube(object):
-    """ Control the tesseract by directly toggling pins using GPIO. """
+    """ Raspberry Pi 1 hardware abstraction layer (HAL) for controlling the tesseract by directly toggling pins using
+    GPIO.
+    """
 
     def __init__(self):
-        pass
+        """ Initialization
+        """
+        self._fps = None
+        self._flattened_frame = None
+        self._tesseract = None
+        self._gpio_setup = False
 
     def setup(self, fps):
+        """ Set up the cube hardware abstraction layer, which means:
+         - instantiate a tesseract interface
+         - set up the GPIO interface
+         - clock in the LED dot-correction data
+        :param fps:
+        :return:
+        """
         faulthandler.enable()
         self._tesseract = GpioTesseract()
         self._tesseract.setup()
+        self._gpio_setup = True
```

Comment:
I would personally skip the `_gpio_setup` setup flag and just do `if self._tesseract is not None` in `teardown`.